artifact/v3/header/header.hpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_ARTIFACT_V3_HEADER_PARSER_HPP
+#define MENDER_ARTIFACT_V3_HEADER_PARSER_HPP
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <unordered_map>
+
+#include <nonstd/optional.hpp>
+
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/io.hpp>
+
+#include <artifact/config.hpp>
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace header {
+
+using namespace std;
+
+using nonstd::nullopt;
+using nonstd::optional;
+
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace error = mender::common::error;
+
+//
+// +---header-info
+//
+
+enum class Payload {
+	RootfsImage,
+	ModuleImage,
+	EmptyPayload,
+};
+
+using PayloadType = pair<Payload, string>;
```

Comment:
Would a `struct PayloadType` make sense here? What is the string for? I think it would be nice to give the two parts names.

Reply:
Also, do we even need the string? Isn't it just a string of the type?

Reply:
> Isn't it just a string of the type?

No, in fact this needs a value. This is because you can write your own update-modules, and hence the string is free-form.

So we provide the `Type`, (ie, Update-Module), and you as the end user provides the `Update-Module:Typename`.

Does it make sense?

Reply:
We could consider a struct however though.

Reply:
I see, but a `struct` would still be nicer, I think. We do it everywhere else, AFAICT.

Reply:
Yup, agreed. Moved to a struct.